### PR TITLE
Enable merge filters optimization.

### LINKED ISSUE #71408: Analyzer does not push down filter condition and use index in View with subquery
**Describe the unexpected behaviour**
When using a View with subquery, the Analyzer does not push down the filtering condition. 

**How to reproduce**
- Reproduced on 24.8 LTS and 24.10 latest release

- https://fiddle.clickhouse.com/2be02245-eb36-4fa0-8dfe-62b8236bbfd7
```
DROP TABLE IF EXISTS alpha;
DROP TABLE IF EXISTS alpha__day;

CREATE TABLE alpha
(
    `ts` DateTime64(6),
    `auid` Int64,
)
ENGINE = MergeTree
ORDER BY (auid, ts)
SETTINGS index_granularity = 1;

CREATE VIEW alpha__day
(
    `ts_date` Date,
    `auid` Int64,
)
AS SELECT
    ts_date,
    auid,
FROM
(
    SELECT
        toDate(ts) AS ts_date,
        auid
    FROM alpha
)
WHERE ts_date <= now() - INTERVAL 1 DAY;

INSERT INTO alpha VALUES (now64() - INTERVAL 3 DAY, 1);
INSERT INTO alpha VALUES (now64() - INTERVAL 3 DAY, 2);
INSERT INTO alpha VALUES (now64() - INTERVAL 3 DAY, 3);

EXPLAIN indexes = 1 SELECT auid FROM alpha__day WHERE auid = 1 settings allow_experimental_analyzer = 1;
EXPLAIN indexes = 1 SELECT auid FROM alpha__day WHERE auid = 1 settings allow_experimental_analyzer = 0;
```

Granules are not filtered when using analyzer (`Granules: 3/3`)
```
-- allow_experimental_analyzer = 1
Expression ((Project names + Projection))
  Filter ((WHERE + (Change column names to column identifiers + (Convert VIEW subquery result to VIEW table structure + (Materialize constants after VIEW subquery + (Project names + (Projection + (Change column names to column identifiers + (Project names + Projection)))))))))
    Expression
      ReadFromMergeTree (default.alpha)
      Indexes:
        PrimaryKey
          Keys: 
            ts
          Condition: (_CAST(toDate(ts)) in (-Inf, 1730600442])
          Parts: 3/3
          Granules: 3/3
-- allow_experimental_analyzer = 0
Expression ((Projection + Before ORDER BY))
  Filter ((WHERE + (Convert VIEW subquery result to VIEW table structure + (Materialize constants after VIEW subquery + (Projection + Before ORDER BY)))))
    Filter ((WHERE + (Projection + Before ORDER BY)))
      Expression
        ReadFromMergeTree (default.alpha)
        Indexes:
          PrimaryKey
            Keys: 
              auid
              ts
            Condition: and((materialize(auid) in [1, 1]), and((auid in [1, 1]), and((_CAST(toDate(ts)) in (-Inf, 1730600442]), (auid in [1, 1]))))
            Parts: 1/3
            Granules: 1/3
```

Working as expected when not using subquery in View.

- https://fiddle.clickhouse.com/37304f26-a736-4505-a3a4-04f3bf4a75e3
```
...
CREATE VIEW alpha__day
(
    `ts_date` Date,
    `auid` Int64,
)
AS SELECT
    toDate(ts) AS ts_date,
    auid
FROM alpha
WHERE ts_date <= now() - INTERVAL 1 DAY;
...
EXPLAIN indexes = 1 SELECT auid FROM alpha__day WHERE auid = 1 settings allow_experimental_analyzer = 1;
```

```
Expression ((Project names + Projection))
  Filter ((WHERE + (Change column names to column identifiers + (Convert VIEW subquery result to VIEW table structure + (Materialize constants after VIEW subquery + (Project names + (Projection + (Change column names to column identifiers + (Project names + Projection)))))))))
    Expression
      ReadFromMergeTree (default.alpha)
      Indexes:
        PrimaryKey
          Keys: 
            auid
            ts
          Condition: and((materialize(auid) in [1, 1]), (_CAST(toDate(ts)) in (-Inf, 1730597277]))
          Parts: 1/3
          Granules: 1/3
```

**Expected behavior**
Analyzer should pushdown condition for View with subquery.

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve `query_plan_merge_filters` optimization. Fixes  #71408

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 4ad8273e5f3d16f5a95220824223800b4a356e26 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71539/4ad8273e5f3d16f5a95220824223800b4a356e26/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by serxa — APPROVED** [review]

File: src/Storages/MergeTree/MergeTreeSplitPrewhereIntoReadSteps.cpp [hunk 1]
```diff
@@ -319,20 +315,22 @@ bool tryBuildPrewhereSteps(PrewhereInfoPtr prewhere_info, const ExpressionAction
             /// 1. AND the last condition with constant True. This is needed to make sure that in the last step filter has UInt8 type
             ///    but contains values other than 0 and 1 (e.g. if it is (number%5) it contains 2,3,4)
             /// 2. CAST the result to the exact type of the PREWHERE column from the original DAG
-            const auto & last_step_result_node_info = node_remap[steps.back().column_name];
+            //const auto & last_step_result_node_info = node_remap[steps.back().column_name];
```
Comment (davenger) [inline src/Storages/MergeTree/MergeTreeSplitPrewhereIntoReadSteps.cpp:318]: remove?

File: src/Storages/MergeTree/MergeTreeIOSettings.h [hunk 2]
```diff
@@ -45,6 +45,8 @@ struct MergeTreeReaderSettings
     bool use_asynchronous_read_from_pool = false;
     /// If PREWHERE has multiple conditions combined with AND, execute them in separate read/filtering steps.
     bool enable_multiple_prewhere_read_steps = false;
+    /// In case of multiple prewhere steps, execute filtering earlier to support short-circuit properly.
+    bool force_shirt_circuit_execution = false;
```
Comment (geldot) [inline src/Storages/MergeTree/MergeTreeIOSettings.h:49]: typo 👕 -> 🩳

**Comment by geldot** [comment]
@KochetovNicolai "shirt" still appears in some of the function param names, worth doing a grep

**Comment by Algunenano** [comment]
Some bugs found by SQLancer. Disabling the setting by default again, but leaving the changes -> https://github.com/ClickHouse/ClickHouse/pull/71890

File: src/Processors/QueryPlan/FilterStep.cpp [hunk 3]
```diff
@@ -76,18 +184,45 @@ void FilterStep::transformPipeline(QueryPipelineBuilder & pipeline, const BuildQ
 void FilterStep::describeActions(FormatSettings & settings) const
 {
     String prefix(settings.offset, settings.indent_char);
+
+    auto cloned_dag = actions_dag.clone();
+
+    std::vector<ActionsAndName> and_atoms;
+    if (!actions_dag.hasStatefulFunctions())
+        and_atoms = splitAndChainIntoMultipleFilters(cloned_dag, filter_column_name);
```
Comment (baibaichen) [inline src/Processors/QueryPlan/FilterStep.cpp:192]: After building pipeline,  `action_dag`  is cleared. calling `describeActions` will failed at `splitAndChainIntoMultipleFilters`

Comment (KochetovNicolai) [reply src/Processors/QueryPlan/FilterStep.cpp:192]: This is true for most other steps as well.
Afiak we really call explain after pipeline is build.
I was planning to make `explain` destructive as well; but it is not convenient while debugging.

File: src/Storages/MergeTree/MergeTreeSplitPrewhereIntoReadSteps.cpp [hunk 4]
```diff
@@ -152,14 +154,12 @@ const ActionsDAG::Node & addFunction(
 const ActionsDAG::Node & addCast(
         const ActionsDAGPtr & dag,
         const ActionsDAG::Node & node_to_cast,
-        const DataTypePtr & to_type,
-        OriginalToNewNodeMap & node_remap)
+        const DataTypePtr & to_type)
 {
     if (!node_to_cast.result_type->equals(*to_type))
```
Comment (geldot) [inline src/Storages/MergeTree/MergeTreeSplitPrewhereIntoReadSteps.cpp:159]: Should be `if (node_to_cast.result_type->equals(*to_type))` (without negation) to fix test `03271_sqllancer_having_issue.sql`? Didn't try, just a hunch.

**Pull request merged**